### PR TITLE
Disable definition-grouping when doing a FindImplementations/GoToDef call.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This PR changes GoToImplementation behavior so that instead of seeing:

https://cloud.githubusercontent.com/assets/4564579/22447059/f84d8cea-e704-11e6-9daa-49565095f54e.png

You instead see: 

image

**Comment by CyrusNajmabadi** [comment]
The way we accomplish this is to actually disable 'grouping by definition' when showing result for FindImplementations/GoToDef.  This is because 'grouping by definition' is exceedingly unhelpful in these scenarios, and just serves to clutter the display by showing two almost identical items per symbol.  The first item is just the SymbolDisplay for the definition, and the second is the source location.  These are generally so close as to be silly to show twice.

Note: disabling 'grouping by definition' has consequences i had to deal with.  Namely, if you disable the grouping, then by default all future usages of the window would have that grouping disabled.  This means that if you did:  "FindRefs, FindImpls, FindRefs" your last FindRefs would not having this grouping, even if that was something you wanted.  

To deal with this, Roslyn now remembers the user's preference for 'group by definition' and restores it as necessary.  i.e. in the same set of calls as above, we will remember the GroupingPreference, then we'll disable it (for FindImpls), then we'l restore it for the last FindRefs.

--

Also, to keep code simple, i ended up breaking up multiple pieces of functionality.  i.e. the 'context object' we pass back for FindRefs has a lot of logic to deal with no references actually being found.  This is not something we need to care about in the FindImpls/GoToDef case (neitehr have 'references', they're only displaying Definitions).  So breaking code out helped keep this special behavior separated.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
@dpoeschl In the case where we get no implementations, you get this message:

image

We don't even go through the Streaming-Presenter codepath.

File: src/VisualStudio/Core/Next/FindReferences/StreamingFindUsagesPresenter.cs [hunk 1]
```diff
@@ -47,34 +51,104 @@ internal partial class StreamingFindUsagesPresenter :
...
+            // We want to still show the column as long as the user has not disabled it themselves.
+            var definitionColumn = window.GetDefinitionColumn();
+            if (definitionColumn.GroupingPriority != desiredGroupingPriority)
+            {
+                SetDefinitionGroupingPriority(window, desiredGroupingPriority);
+            }
+
+            // If the user changes the grouping, then store their current preference.
+            var tableControl = (IWpfTableControl2)window.TableControl;
+            tableControl.GroupingsChanged += (s, e) => StoreCurrentGroupingPriority(window);
```
Comment (tmeschter) [inline src/VisualStudio/Core/Next/FindReferences/StreamingFindUsagesPresenter.cs]: Isn't this going to create and hook up a new delegate every time the method is called?

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/FindReferences/StreamingFindUsagesPresenter.cs]: Not quite.  The call to ```var window = _vsFindAllReferencesService.StartSearch(title);``` will give us a window with a fresh table control.

**Pull request merged**